src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,309 @@
+    try:
+        return cast(
+            NetworkSecurityGroup,
+            network_client.network_security_groups.get(resource_group, name),
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        logging.debug("nsg %s does not exist %s", name, err)
+        return None
+
+
+def create_nsg(name: str, location: Region) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+
+    logging.info("creating nsg %s:%s:%s", resource_group, location, name)
+    network_client = get_network_client()
+
+    params: Dict = {
+        "location": location,
+    }
+
+    if "ONEFUZZ_OWNER" in os.environ:
+        params["tags"] = {"OWNER": os.environ["ONEFUZZ_OWNER"]}
+    try:
+        network_client.network_security_groups.begin_create_or_update(
+            resource_group, name, params
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if "The request failed due to conflict with a concurrent request" in str(err):
+            logging.debug(
+                "create NSG had conflicts with concurrent request, ignoring %s", err
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_CREATE,
+            errors=["Unable to create nsg %s due to %s" % (name, err)],
+        )
+    return None
+
+
+def update_nsg(nsg: NetworkSecurityGroup) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+
+    logging.info("updating nsg %s:%s:%s", resource_group, nsg.location, nsg.name)
+    network_client = get_network_client()
+
+    try:
+        network_client.network_security_groups.begin_create_or_update(
+            resource_group, nsg.name, nsg
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if "The request failed due to conflict with a concurrent request" in str(err):
+            logging.debug(
+                "create NSG had conflicts with concurrent request, ignoring %s", err
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_CREATE,
+            errors=["Unable to create nsg %s due to %s" % (nsg.name, err)],
+        )
+    return None
+
+
+def delete_nsg(name: str) -> Any:
+    resource_group = get_base_resource_group()
+
+    logging.info("deleting nsg: %s %s", resource_group, name)
+    network_client = get_network_client()
+    return network_client.network_security_groups.begin_delete(resource_group, name)
+
+
+def set_allowed(name: str, sources: List[str]) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot update nsg rules. nsg %s not found" % name],
+        )
+
+    logging.info(
+        "setting allowed incoming connection sources for nsg: %s %s",
+        resource_group,
+        name,
+    )
+    security_rules = []
+    min_priority = 100
+    max_priority = 4096
```

Comment:
Is this an inclusive upper bound?

If so, and the set of priorities is `{100, 101, ..., 4095, 4096}`, then don't we want `max_rule_count = max_priority - min_priority + 1`?

Also, I suggest computing `max_rule_count` explicitly as a local var, and using it both in the check and in the error message.

Ideally we'd also do this input sanitization earlier in the function, before making multiple Azure calls.

Reply:
Actually there are only 1000 rules are allowed, so all that logic is a bit different